src/gui/guiFormSpecMenu.cpp

```diff
@@ -1679,22 +1679,60 @@ void GUIFormSpecMenu::parseHyperText(parserData *data, const std::string &elemen
 
 void GUIFormSpecMenu::parseLabel(parserData* data, const std::string &element)
 {
-	std::vector<std::string> parts = split(element,';');
+	std::vector<std::string> parts = split(element, ';');
 
 	if ((parts.size() == 2) ||
 		((parts.size() > 2) && (m_formspec_version > FORMSPEC_API_VERSION)))
 	{
-		std::vector<std::string> v_pos = split(parts[0],',');
+		std::vector<std::string> v_pos = split(parts[0], ',');
 		std::string text = parts[1];
 
-		MY_CHECKPOS("label",0);
+		MY_CHECKPOS("label", 0);
 
-		if(!data->explicit_size)
-			warningstream<<"invalid use of label without a size[] element"<<std::endl;
+		if (!data->explicit_size)
+			warningstream << "Invalid use of label without a size[] element." << std::endl;
 
 		std::vector<std::string> lines = split(text, '\n');
+		auto style = getStyleForElement("label", "");
+
+		EGUI_ALIGNMENT horigin = gui::EGUIA_LOWERRIGHT;
+		EGUI_ALIGNMENT vorigin = gui::EGUIA_SCALE; // Used as base
+		std::string str_horigin = trim(style.get(StyleSpec::HORIGIN, ""));
+		std::string str_vorigin = trim(style.get(StyleSpec::VORIGIN, ""));
+
+		if (str_horigin == "left")
+			horigin = gui::EGUIA_UPPERLEFT;
+		else if (str_horigin == "center")
+			horigin = gui::EGUIA_CENTER;
+		else if (str_horigin == "base")
+			horigin = gui::EGUIA_SCALE;
+
+		if (str_vorigin == "top")
+			vorigin = gui::EGUIA_UPPERLEFT;
+		else if (str_vorigin == "center")
+			vorigin = gui::EGUIA_CENTER;
+		else if (str_vorigin == "bottom")
+			vorigin = gui::EGUIA_LOWERRIGHT;
+
+		s32 font_height = font_line_height(m_font);
+
+		std::string str_newline = style.get(StyleSpec::NEWLINE_SPACING, "0.5");
+		float newline_spacing;
+		if (trim(str_newline) == "auto")
+			newline_spacing = (float)font_height;
+		else
+			newline_spacing = stof(str_newline) * (float)imgsize.Y;
+
+		s32 rect_middle = newline_spacing * (float)lines.size() / 2;
 
-		for (unsigned int i = 0; i != lines.size(); i++) {
+		s32 first_char_base = 0;
+		if (horigin == gui::EGUIA_SCALE && lines[0].size() > 0) {
+			std::wstring unescaped = unescape_enriched(translate_string(
+				utf8_to_wide(unescape_string(lines[0]))));
+			first_char_base = m_font->getDimension(std::string(unescaped[0], 1)).Width / 2;
```

Comment:
I'm getting compile errors here:

```
/home/ruben/dev/minetest/src/gui/guiFormSpecMenu.cpp: In member function ‘void GUIFormSpecMenu::parseLabel(GUIFormSpecMenu::parserData*, const string&)’:
/home/ruben/dev/minetest/src/gui/guiFormSpecMenu.cpp:1732:48: error: cannot convert ‘std::string’ {aka ‘std::__cxx11::basic_string<char>’} to ‘const wchar_t*’
 1732 |    first_char_base = m_font->getDimension(std::string(unescaped[0], 1)).Width / 2;
      |                                                ^~~~~~~~~~~~~~~~~~~~~~~
      |                                                |
      |                                                std::string {aka std::__cxx11::basic_string<char>}
In file included from /home/ruben/dev/minetest/src/gui/guiFormSpecMenu.cpp:38:
/usr/include/irrlicht/IGUIFont.h:58:61: note:   initializing argument 1 of ‘virtual irr::core::dimension2d<unsigned int> irr::gui::IGUIFont::getDimension(const wchar_t*) const’
   58 |  virtual core::dimension2d<u32> getDimension(const wchar_t* text) const = 0;
      |                                              ~~~~~~~~~~~~~~~^~~~
/home/ruben/dev/minetest/src/gui/guiFormSpecMenu.cpp: In member function ‘void GUIFormSpecMenu::parseVertLabel(GUIFormSpecMenu::parserData*, const string&)’:
/home/ruben/dev/minetest/src/gui/guiFormSpecMenu.cpp:1866:33: error: invalid initialization of reference of type ‘const string&’ {aka ‘const std::__cxx11::basic_string<char>&’} from expression of type ‘std::__cxx11::basic_string<wchar_t>’
 1866 |     utf8_to_wide(unescape_string(lines[0]))));
      |                  ~~~~~~~~~~~~~~~^~~~~~~~~~
In file included from /home/ruben/dev/minetest/src/metadata.h:25,
                 from /home/ruben/dev/minetest/src/itemstackmetadata.h:22,
                 from /home/ruben/dev/minetest/src/inventory.h:24,
                 from /home/ruben/dev/minetest/src/inventorymanager.h:22,
                 from /home/ruben/dev/minetest/src/gui/guiFormSpecMenu.h:27,
                 from /home/ruben/dev/minetest/src/gui/guiFormSpecMenu.cpp:26:
/home/ruben/dev/minetest/src/util/string.h:67:46: note: in passing argument 1 of ‘std::wstring utf8_to_wide(const string&)’
   67 | std::wstring utf8_to_wide(const std::string &input);
      |                           ~~~~~~~~~~~~~~~~~~~^~~~~
/home/ruben/dev/minetest/src/gui/guiFormSpecMenu.cpp:1874:51: error: no matching function for call to ‘max(float&, unsigned int)’
 1874 |       m_font->getDimension(lines[i].c_str()).Width);
```

Reply:
I'm aware of the compiler errors; due to the weird and unique way I compile Minetest, I have to push to GitHub before I can compile.  I'll notify you when I have it working.